backend/src/main/java/fk/prof/backend/model/policy/ZookeeperBasedPolicyStoreAPI.java

```diff
@@ -0,0 +1,232 @@
+        return appIds.stream().filter(s -> s.startsWith(prefix)).collect(Collectors.toSet());
+
+    }
+
+    Set<String> getClusterIds(String appId, String prefix) {
+        if (prefix == null) return new HashSet<>();
+        Set<String> clusterIds = inMemoryPolicyCache.getClusterIds(appId);
+        return clusterIds.stream().filter(s -> s.startsWith(prefix)).collect(Collectors.toSet());
+    }
+
+    Set<String> getProcNames(String appId, String clusterId, String prefix) {
+        if (prefix == null) return new HashSet<>();
+        Set<String> procNames = inMemoryPolicyCache.getProcNames(appId, clusterId);
+        return procNames.stream().filter(s -> s.startsWith(prefix)).collect(Collectors.toSet());
+    }
+
+    @Override
+    public PolicyDTO.PolicyDetails getPolicy(Recorder.ProcessGroup processGroup) {
+        if (processGroup == null) {
+            logger.info("Process Group is null");
+            return null;
+        }
+        return inMemoryPolicyCache.get(processGroup.getAppId(), processGroup.getCluster(), processGroup.getProcName());
+    }
+
+    @Override
+    public Future<Void> createPolicy(Recorder.ProcessGroup processGroup, PolicyDTO.PolicyDetails policyDetails) {
+        Future<Void> future = Future.future();
+        if (processGroup == null || policyDetails == null) {
+            future.fail("One of processGroup(" + processGroup + ") or policyDetails(" + policyDetails + ") is null");
+            return future;
+        }
+        if (getPolicy(processGroup) != null) {
+            future.fail("Policy already exists");
+            return future;
+        }
+        setPolicy(future, processGroup, policyDetails);
+        return future;
+    }
+
+    @Override
+    public Future<Void> updatePolicy(Recorder.ProcessGroup processGroup, PolicyDTO.PolicyDetails policyDetails) {
+        Future<Void> future = Future.future();
+        if (processGroup == null || policyDetails == null) {
+            future.fail("One of processGroup(" + processGroup + ") or policyDetails(" + policyDetails + ") is null");
+            return future;
+        }
+        if (getPolicy(processGroup) == null) {
+            future.fail("Policy does not exists ");
+            return future;
+        }
+        if (getPolicy(processGroup).equals(policyDetails)) {
+            future.fail("Nothing to change in policy");
+            return future;
+        }
+        setPolicy(future, processGroup, policyDetails);
+        return future;
+    }
+
+    private void setPolicy(Future<Void> future, Recorder.ProcessGroup processGroup, PolicyDTO.PolicyDetails
+            policyDetails) {
+        String zNodePath = policyPath + DELIMITER + encode(processGroup.getAppId()) + DELIMITER + encode(processGroup.getCluster()) + DELIMITER + encode(processGroup.getProcName()) + DELIMITER + POLICY_NODE_PREFIX;
+        boolean acquired;
+        try {
+            acquired = setPolicySemaphore.tryAcquire(1, TimeUnit.SECONDS);
```

Comment:
why take a lock across all policies. the lock should be specific to a process group, right? Set policy on different process groups should be allowed to proceed in parallel